File: Examples/OSX/AudioKitDemo/AudioKitDemo/PlayFile/AudioFilePlayer.m
//
//  AudioFilePlayer.m
//  AudioKit Example
//
//  Created by Aurelius Prochazka on 6/16/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "AudioFilePlayer.h"

@implementation AudioFilePlayer

- (instancetype)init {
    self = [super init];
    if (self) {
        
        // NOTE BASED CONTROL ==================================================
        AudioFilePlayerNote *note = [[AudioFilePlayerNote alloc] init];
        [self addNoteProperty:note.speed];
        
        // INSTRUMENT DEFINITION ===============================================
        NSString *file;
        file = [[NSBundle mainBundle] pathForResource:@"blaster" ofType:@"aiff"];
        AKSoundFile *fileTable;
        fileTable = [[AKSoundFile alloc] initWithFilename:file];
        [self addFunctionTable:fileTable];
        
        AKMonoSoundFileLooper *looper;
        looper = [AKMonoSoundFileLooper looperWithSoundFile:fileTable];
        looper.frequencyRatio = note.speed;
        looper.loopMode = [AKMonoSoundFileLooper loopPlaysOnce];
        [self connect:looper];
        
        AKReverb *reverb;
        reverb = [AKReverb reverbWithInput:looper];
        reverb.feedback.value = 0.85;
        [self connect:reverb];
        
        // AUDIO OUTPUT ========================================================
        AKAudioOutput *audio;
        audio = [[AKAudioOutput alloc] initWithStereoAudioSource:reverb];
        [self connect:audio];
    }
    return self;
}

@end

// -----------------------------------------------------------------------------
#  pragma mark - Instrument Note
// -----------------------------------------------------------------------------

@implementation AudioFilePlayerNote

- (instancetype)init;
{
    self = [super init];
    if(self) {
        _speed = [[AKNoteProperty alloc] initWithValue:1.0
                                               minimum:0.2
                                               maximum:2.0];
        [self addProperty:_speed];
    }
    return self;
}
- (instancetype)initWithSpeed:(float)speed;
{
    self = [self init];
    if(self) {
        self.speed.value = speed;
    }
    return self;
}

@end

File: Examples/iOS/Swift/AudioKitDemo/AudioKitDemo/AudioFilePlayer.swift
//
//  AudioFilePlayer.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//


class AudioFilePlayer: AKInstrument {
    
    var auxilliaryOutput = AKAudio()

    // INSTRUMENT BASED CONTROL ============================================
    var speed       = AKInstrumentProperty(value: 1, minimum: -2, maximum: -2)
    var scaling     = AKInstrumentProperty(value: 1, minimum: 0.0, maximum: 3.0)
    var sampleMix   = AKInstrumentProperty(value: 0, minimum: 0, maximum: 1)
    
    
    // INSTRUMENT DEFINITION ===============================================
    override init() {
        super.init()
        
        addProperty(speed)
        addProperty(scaling)
        addProperty(sampleMix)
        
        let file1 = String(NSBundle.mainBundle().pathForResource("PianoBassDrumLoop", ofType: "wav")!)
        let file2 = String(NSBundle.mainBundle().pathForResource("PianoBassDrumLoop", ofType: "wav")!)
        
        let fileIn1 = AKFileInput(filename: file1)
        fileIn1.speed = speed;
        connect(fileIn1)
        
        let fileIn2 = AKFileInput(filename: file2)
        fileIn2.speed = speed;
        connect(fileIn2)

        var fileInLeft  = AKMix(input1: fileIn1.leftOutput,  input2: fileIn2.leftOutput, balance: sampleMix)
        connect(fileInLeft)
        var fileInRight = AKMix(input1: fileIn1.rightOutput, input2: fileIn2.rightOutput, balance: sampleMix)
        connect(fileInRight)

        var leftF = AKFFT(
            input: fileInLeft.scaledBy(0.25.ak),
            fftSize: (1024.ak),
            overlap: (256.ak),
            windowType: AKFFTWindowType.Hamming,
            windowFilterSize: (1024.ak)
        )
        connect(leftF)
        
        var leftR = AKFFT(
            input: fileInRight.scaledBy(0.25.ak),
            fftSize: (1024.ak),
            overlap: (256.ak),
            windowType: AKFFTWindowType.Hamming,
            windowFilterSize: (1024.ak)
        )
        connect(leftR)

        
        var scaledLeftF = AKScaledFFT(
            signal: leftF,
            frequencyRatio: scaling
        )
        connect(scaledLeftF)
        
        var scaledLeftR = AKScaledFFT(
            signal: leftR,
            frequencyRatio: scaling
        )
        connect(scaledLeftR)

        var scaledLeft  = AKResynthesizedAudio(signal: scaledLeftF)
        var scaledRight = AKResynthesizedAudio(signal: scaledLeftR)
        
        var mono = AKMix(input1: scaledLeft, input2: scaledRight, balance: 0.5.ak)
        
        // Output to global effects processing
        auxilliaryOutput = AKAudio.globalParameter()
        assignOutput(auxilliaryOutput, to: mono)
    }
}